File: atcoder/abc/python/abc219/abc219_F.py

import sys
infile = sys.stdin.buffer

def gs()  : return infile.readline().rstrip()
def gi()  : return int(gs())
def gf()  : return float(gs())
def gss() : return gs().split()
def gis() : return [int(x) for x in gss()]
def gfs() : return [float(x) for x in gss()]

def main(infn="") :
    global infile
    infile = open(infn,"r") if infn else open(sys.argv[1],"r") if len(sys.argv) > 1 else sys.stdin
    S = gs(); K = gi()
    x,y = 0,0; pts = set(); pts.add((x,y))
    for c in S :
        if c == 'L'   : x-=1
        elif c == 'R' : x+=1
        elif c == 'D' : y-=1
        elif c == 'U' : y+=1
        pts.add((x,y))
    dx,dy,ans = x,y,0
    if dx == 0 and dy == 0 :
        ans = len(pts)
    else :
        if dx == 0 :
            pts2 = set()
            for (x,y) in pts : pts2.add((y,x))
            pts = pts2; dx,dy = dy,dx
        if dx < 0 :
            pts2 = set()
            for (x,y) in pts : pts2.add((-x,y))
            pts = pts2; dx,dy = -dx,dy
        eqclass = {}
        for (x,y) in pts :
            if x >= dx : 
                n = x // dx; x2 = x-dx*n; y2 = y-dy*n
            elif x < 0 :
                n = (-x+dx-1)//dx; x2 = x+n*dx; y2 = y+n*dy
            else :
                x2,y2 = x,y
            if (x2,y2) not in eqclass : eqclass[(x2,y2)] = []
            eqclass[(x2,y2)].append((x,y))
        for (x,y) in eqclass :
            ll = eqclass[(x,y)]
            ll.sort()
            for i in range(len(ll)-1) :
                ans += min(K,(ll[i+1][0]-ll[i][0])//dx)
            ans += K
    print(ans)

if __name__ == '__main__' :
    main()
    sys.stdout.flush()

File: abc/go/abc219/abc219_F/abc219_F.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"sort"
	"strconv"
)

type scanner struct { sc *bufio.Scanner }
func newScanner(input io.Reader) *scanner {
	sc := bufio.NewScanner(input)
	sc.Split(bufio.ScanWords)
	sc.Buffer(make([]byte, 1024), int(1e+9))
	return &scanner{sc}	
}
var rdr = newScanner(os.Stdin)
const BUFSIZE = 10000000
var wrtr = bufio.NewWriterSize(os.Stdout, BUFSIZE)
func gs() string  { rdr.sc.Scan(); return rdr.sc.Text() }
func gi() int     { i,e := strconv.Atoi(gs()); if e != nil {panic(e)}; return i }
func min(a,b int) int { if a > b { return b }; return a }

type PI struct { x,y int }

func doInitialTraversal(S string) (map[PI]bool,int,int) {
	pos := PI{0,0}; x := make(map[PI]bool); x[pos] = true
	for _,c := range S { 
		if c == 'L' { pos.x-- }
		if c == 'R' { pos.x++ }
		if c == 'U' { pos.y++ }
		if c == 'D' { pos.y-- }
		x[pos] = true
	}
	return x,pos.x,pos.y
}

func flipxy(oldpmap map[PI]bool,dx,dy int) (map[PI]bool,int,int) {
	pmap := make(map[PI]bool)
	for k := range oldpmap { pmap[PI{k.y,k.x}] = true }
	return pmap,dy,dx
}

func invertx(oldpmap map[PI]bool,dx,dy int) (map[PI]bool,int,int) {
	pmap := make(map[PI]bool)
	for k := range oldpmap { pmap[PI{-k.x,k.y}] = true }
	return pmap,-dx,dy
}

func groupPoints(pmap map[PI]bool, dx int, dy int) ([][]PI) {
	gmap := make(map[PI][]PI)
	for pt := range pmap {
		pt2 := pt
		if pt2.x >= dx { 
			d := pt2.x/dx; pt2.x -= d*dx; pt2.y -= d*dy
		} else if pt2.x < 0 {
			d := (-pt2.x+dx-1)/dx; pt2.x += d*dx; pt2.y += d*dy
		}
		_,ok := gmap[pt2]; if !ok { gmap[pt2] = make([]PI,0) }
		gmap[pt2] = append(gmap[pt2],pt)
	}
	res := make([][]PI,0)
	for _,v := range gmap {	res = append(res,v)	}
	return res
}

func solve(S string, K int) int {
	pmap,dx,dy := doInitialTraversal(S)
	if dx == 0 && dy == 0 || K == 1 { return len(pmap) }
	if dx == 0 { pmap,dx,dy = flipxy(pmap,dx,dy) }
	if dx < 0 { pmap,dx,dy = invertx(pmap,dx,dy) }
	ptgroups := groupPoints(pmap,dx,dy)
	ans := 0
	for _,ptg := range ptgroups {
		sort.Slice(ptg,func(i,j int)bool{return ptg[i].x < ptg[j].x})
		for i:=0;i<len(ptg)-1;i++ {
			ans += min(K,(ptg[i+1].x-ptg[i].x)/dx)
		}
		ans += K // furthest point should get full marks
	}
	return ans
}

func main() {
	//f1, _ := os.Create("cpu.prof"); pprof.StartCPUProfile(f1); defer pprof.StopCPUProfile()
    defer wrtr.Flush()
	infn := ""
	if infn == "" && len(os.Args) > 1 {	infn = os.Args[1] }
    if infn != "" {	f, e := os.Open(infn); if e != nil { panic(e) }; rdr = newScanner(f) }
	S := gs(); K := gi()
	ans := solve(S,K) 
	fmt.Println(ans)
}
